File: 8kyu/Reverse-Word.java
/*

        8 kyu
        Reversed Words
        54116792% of 6,85884 of 64,941jhoffner2 Issues Reported
        Java
        17
        VIM
        EMACS
        Instructions
        Output
        Complete the solution so that it reverses all of the words within the string passed in.

        Words are separated by exactly one space and there are no leading or trailing spaces.

        Example(Input --> Output):

        "The greatest victory is that which requires no battle" --> "battle no requires which that is victory greatest The"

*/

public class ReverseWords{
    public static String reverseWords(String str){
        String[] words = str.split(" ");
        StringBuilder sb = new StringBuilder();
        for (int i = words.length - 1; i >= 0; i--) {
            sb.append(words[i]).append(" ");
        }
        return sb.toString().trim();
    }
}

/*
TEST RESULTS:
        Time: 2584ms Passed: 2Failed: 0
        Test Results:
        SolutionTest
        randomTests()
        fixedTests()
        Completed in 68ms
        You have passed all of the tests! :)
*/
/*
 * ANOTHER SOLUTION
 */
public class ReverseWords{
    public static String reverseWords(String str){
        char[] s = str.toCharArray();
        int i = 0, j = 0;
        while (j <= str.length()) {
            if (j == str.length() || s[j] == ' ') {
                reverse(s, i, j - 1);
                i = j + 1;
            }
            j++;
        }
        reverse(s, 0, str.length() - 1);
        return new String(s);
    }
    private static void reverse(char[] s, int start, int end) {
        while (start < end) {
            char temp = s[start];
            s[start] = s[end];
            s[end] = temp;
            start++;
            end--;
        }
    }
}

File: 7kyu/Reverse-word.go
/*
		7 kyu
		Reverse words
		159737791% of 7,8881,224 of 94,048jnicol3 Issues Reported
		Go
		1.18
		VIM
		EMACS
		Instructions
		Output
		Complete the function that accepts a string parameter, and reverses each word in the string. All spaces in the string should be retained.

		Examples
		"This is an example!" ==> "sihT si na !elpmaxe"
		"double  spaces"      ==> "elbuod  secaps"
*/
package kata

import "strings"

func ReverseWords(str string) string {
	words := strings.Split(str, " ")
	for i, word := range words {
		words[i] = reverse(word)
	}
	return strings.Join(words, " ")
}

func reverse(s string) string {
	r := []rune(s)
	for i, j := 0, len(r)-1; i < len(r)/2; i, j = i+1, j-1 {
		r[i], r[j] = r[j], r[i]
	}
	return string(r)
}
/*
TEST  RESULTS:
		Time: 2059ms Passed: 2Failed: 0
		Test Results:
		Test Example
		should work for sample test cases
		should work for random tests
		You have passed all of the tests! :)
*/